### PR TITLE
fix potential NPE

### LINKED ISSUE #29: NPE Crash with Invalid Credentials in ClientPool::GuardedClient::TestConnection
## Problem

When connecting with empty or invalid credentials, the client can crash with a NullPointerException.

### Issue 1: NPE in `TestConnection()` with nullptr client

**File:** `timeplus/timeplus.cpp` lines 74, 78, 96

**Root Cause:**
1. `GuardedClient` default constructs with `client = nullptr`
2. If `GetGuardedClient()` throws (e.g., `TimeoutError` when pool is exhausted), `client` remains `nullptr`
3. The catch blocks call `guarded_client.TestConnection()` with `client == nullptr`
4. `TestConnection()` calls `client->Ping()` causing NPE

```cpp
void ExecuteWithRetries(std::function<void(Client&)> func, BaseResult& result) {
    ClientPool::GuardedClient guarded_client;  // client = nullptr
    for (int retries = config_.max_retries; retries >= 0; --retries) {
        try {
            if (guarded_client.client == nullptr) {
                guarded_client = client_pool_.GetGuardedClient(...);  // May throw TimeoutError
            }
            ...
        } catch (const ProtocolError& ex) {
            guarded_client.TestConnection();  // NPE if client is nullptr!
        } catch (const ServerError& ex) {
            guarded_client.TestConnection();  // NPE if client is nullptr!
        } catch (const std::exception& ex) {
            guarded_client.TestConnection();  // NPE if client is nullptr!
        }
    }
}
```

**Fix for Issue 1:**

Add null check in `TestConnection()` (`timeplus/client_pool.cpp`):

```cpp
void ClientPool::GuardedClient::TestConnection() noexcept {
    if (!client) {
        valid = false;
        return;
    }
    try {
        client->Ping();
        valid = true;
    } catch (...) {
        valid = false;
    }
    ...
}
```

### Issue 2: NPE in TRACE macro (debug builds only)

**File:** `timeplus/client_pool.cpp` line 40

`GetCurrentEndpoint()` returns `std::optional<Endpoint>`, but the TRACE macro dereferenced it without checking. This only affects debug builds with `TRACE_TIMEPLUS_CPP` defined.

**Note:** This was already fixed by adding a null check before the TRACE statement.

## Files to Modify

- `timeplus/client_pool.cpp` - Add null check for `client` in `TestConnection()`

## Verification

1. Build the project
2. Test with invalid credentials or exhausted connection pool to verify no crash occurs

### PR DESCRIPTION
**Pull request by yuzifeng1984**
Fix #29

### CODE REVIEW

**Review by yokofly — COMMENTED** [review]
P2 valid. `ClientPool` tracks queue slots, not only live `Client` objects, so `GuardedClient::~GuardedClient()` still needs to call `Release(std::move(client), valid)` whenever `pool_` is set, even if `client == nullptr`. Skipping that permanently drops a slot, which can lead to acquire timeouts and the pool-size assert on destruction. Minimal fix: restore `if (pool_)` here; the same cleanup rule likely also applies to the move-assignment path.

File: timeplus/client_pool.h [hunk 1]
```diff
@@ -36,7 +36,7 @@ class ClientPool {
         GuardedClient(ClientPool* pool, ClientPtr client, bool valid) : client(std::move(client)), valid(valid), pool_(pool) {}
 
         ~GuardedClient() {
-            if (pool_) {
+            if (pool_ && client) {
```
Comment (yokofly) [inline timeplus/client_pool.h]: That is a pool-accounting bug, not just a style issue. ClientPool is managing queue slots, not only live Client instances. The pool is initially filled with (nullptr, false) placeho
  lders, and Acquire() already treats nullptr as a valid empty slot to lazily initialize later. If the destructor skips Release() when client == nullptr, one slot disappears permanent
  ly. After enough of those, Acquire() can time out and the pool destructor assertion can fail. The review comment matches that behavior.  
  Actionable fix:

  - Revert the destructor guard in timeplus/client_pool.h to check only pool_.
  - Keep the TestConnection() null check from the PR; that part is correct.
  - Add a regression test for “move/reset client to null, then guard destruction still returns the slot”.

  One more thing: the same accounting assumption still exists in GuardedClient::operator=, which currently releases only when pool_ && client. That is the same latent bug pattern, just
  outside the reviewed hunk. I would either fix it in the same PR or at least leave a follow-up note, because otherwise the destructor path is fixed but reassignment can still drop
  slots.

Comment (yuzifeng1984) [reply timeplus/client_pool.h]: The client is inited in ClientPool::Acquire so should not be nullptr. Guard an nullptr client may be useless.

Comment (yokofly) [reply timeplus/client_pool.h]: The issue is after `Acquire()`, not inside it. `guarded.client` can later become `nullptr` while `pool_` is still set, e.g. after `auto client = std::move(guarded.client);` or `guarded.client.reset();`. 


In that case the destructor still needs to return the empty slot via `Release(std::move(client), valid)`, otherwise the pool permanently loses capacity.

Comment (yuzifeng1984) [reply timeplus/client_pool.h]: It is invalid to change / move client out of guard unless the caller will take the responsibility to release it. Any code do this? Anyway,  this is not the reason we put back nullptr to the pool.

Comment (yokofly) [reply timeplus/client_pool.h]: Even if moving `client` out of the guard is treated as invalid usage, the pool-accounting issue is independent of that. This pool uses `nullptr` as the empty-slot sentinel: it is initialized with `(nullptr, false)`, and `Acquire()` lazily materializes a client from that slot. So `Release(nullptr, valid)` is still a valid way to return the slot. The destructor should keep the `pool_` check only, so slot count is preserved.


So the fix here is to restore `if (pool_)` instead of `if (pool_ && client)`.

Comment (yokofly) [reply timeplus/client_pool.h]: Safe case:

  std::unique_ptr<Client> p = nullptr;
  pool_->Release(std::move(p), valid);

  What moves is the unique_ptr object, not some “uncontrolled null”. The destination just receives an empty unique_ptr.

###  Why the client check is wrong:

  Pool size = 2

  Initial queue:
  [ slot A: (nullptr, false) ]
  [ slot B: (nullptr, false) ]
  queue size = 2

  Acquire() takes one slot out:

  After Acquire():
  queue:
  [ slot B: (nullptr, false) ]
  queue size = 1

  guard holds:
  pool_  -> ClientPool
  client -> unique_ptr<Client>   // maybe later becomes null
  valid  -> true/false

###  If destructor does if (pool_) Release(...):

  guard destroyed
  Release(nullptr, valid)   // still returns the slot token

  queue:
  [ slot B: (nullptr, false) ]
  [ slot A: (nullptr, valid) ]
  queue size = 2   <- correct

###  If destructor does if (pool_ && client) Release(...) and client == nullptr:

  guard destroyed
  skip Release()

  queue:
  [ slot B: (nullptr, false) ]
  queue size = 1   <- slot leaked

  After leaking pool_size such slots:

  queue size = 0
  Acquire() blocks / times out
  ~ClientPool() assert(clients_.size() == pool_size_) can fail

  So the reason to avoid the client check is simple: this queue is tracking capacity tokens, not only live Client instances. An empty unique_ptr is a valid returned slot.

Comment (yuzifeng1984) [reply timeplus/client_pool.h]: 1. If client is acquired, it will be released in dtor.
2. If client is not acquired in Acquire(), pool will not lost slot so no need to release it.
3. Add back nullptr in dtor wrongly increase the slot count though pool_ is most likely (if not always) null in such case.

**Review by yokofly — APPROVED** [review]
